README.md

```diff
@@ -1,2 +1,422 @@
+
+## Features
+
+✅ Compile-time safe
+
+✅ Thread safe
+
+✅ Hierarchical dependency scoping
+
+✅ Constructor injection
+
+✅ Multi-module support
+
+✅ Dependency inversion support
+
+✅ Transitive dependency solving
+
+✅ Cycle detection
+
+✅ Architecture independent
+
+✅ Simple integration: no DI-specific types or generics required
+
+✅ Easy testing: every type has a memberwise initializer
+
+✅ Clear error messages: never debug generated code
+
+## Using SafeDI
+
+SafeDI utilizes Swift Macros and Swift Package Manager plugins to read your code and generate a dependency tree that is validated at compile time. Dependencies can either be instantiated by SafeDI or forwarded into the SafeDI dependency tree.
+
+Opting a type into the SafeDI dependency tree is straightforward: add the `@Instantiable` macro to your type declaration, and decorate your type‘s dependencies with macros that signal the lifecycle of each property. Decorate a property with `@Instantiated` if you want to initialize it when the enclosing type is initialized; `@Forwarded` if you want to propagate a runtime-determine value down the dependency tree; or `@Received` if you want to receive the property from an `@Instantiated` or `@Forwarded` property further up the dependency tree.
+
+If a type is declared in third-party code, you can declare an extension the type in your code and decorate it with the `@ExternalInstantiable` macro to opt it into SafeDI.
+
+Let‘s walk through each of these macros in detail.
+
+### @Instantiable
+
+Type declarations decorated with the [`@Instantiable` macro](Sources/SafeDI/PropertyDecoration/Instantiable.swift) are able to be instantiated by SafeDI. Types decorated with this macro can instantiate other `@Instantiable` dependencies, forward dependencies injected from outside of SafeDI, or receive dependencies instantiated or forwarded by objects further up the dependency tree.
+
+Every `@Instantiable`-decorated type must be:
+
+1. `public` or `open`
+
+2. Have a `public init(…)` or `open init(…)` method that receives every injectable property
+
+The `@Instantiable` guides engineers through satisfying these requirements with build-time FixIts.
+
+#### Example
+
+Here is a sample `UserService` implementation that is `@Instantiable`:
+
+```swift
+import SafeDI
+
+/// A protocol that defines a UserService.
+/// It is not necessary to utilize protocols with SafeDI, but since
+/// protocol-driven development aids both testability and dependency
+/// inversion, our examples show how protocols can be used with SafeDI.
+public protocol UserService {
+    var user: User? { get }
+    func login(username: String, password: String) async throws -> User
+}
+
+/// A default implementation of `UserService` that can fulfill `@Instantiated`
+/// properties of type `UserService` or `DefaultUserService`.
+@Instantiable(fulfillingAdditionalTypes: [UserService.self])
+public final class DefaultUserService: UserService {
```

Comment:
I know we talked about this probably about 100 times, and I know you disagree due to differences in coding style across the industry, but seeing this in front of me really does make me want to say this again 😂 

I really do think we should be able to infer that `XYZProtocolName: ProtocolName` fulfills the type `ProtocolName`, _unless_ `fulfillingAdditionalTypes` is provided as an argument.

Otherwise we're writing effectively boilerplate (We're writing the string `UserService` 3 times), which feels like something Macros should eliminate.

Reply:
We disagree on the point of a macro! The point isn't eliminating boilerplate as much as making it easier to do a standard thing.

At a high level: DI systems are already quite magical, so I really want to avoid "we made this assumption, didn't have a good way to tell you that we did so, and gave you no way to opt out without changing your type name" scenarios.

Imagine we had:
```
public final class ViewController: UIViewController { ... }

@Instantiable
public final class ProfileViewController: ViewController { ... }
```

I kinda doubt we would want to make `ProfileViewController` also fulfill `ViewController`. And if `ViewController` were `@Instantiable` the decision tree gets even more complex. Of course, we could come up with some set of rules to handle these cases and document them here, but... this kind of magic would make the learning curve steeper for this library.

The learning curve for adopting automated DI is already kinda reasonably high. I do not want to make it more complicated when the alternative is "repeat yourself in a way that makes it very clear what you're doing".